File: prefig/core/rectangle.py
## Add a graphical element describing a rectangle

import lxml.etree as ET
import logging
from . import user_namespace as un
from . import utilities as util
from . import math_utilities as math_util
from . import CTM

log = logging.getLogger('prefigure')

# Process a rectangle tag
def rectangle(element, diagram, parent, outline_status):
    if outline_status == 'finish_outline':
        finish_outline(element, diagram, parent)
        return

    # An author may specifiy either the lower-left corner or the center
    try:
        ll = un.valid_eval(element.get('lower-left', '(0,0)'))
        dims = un.valid_eval(element.get('dimensions', '(1,1)'))
        center = element.get('center', None)
        if center is not None:
            center = un.valid_eval(center)
            ll = center - 0.5 * dims
        else:
            center = ll + 0.5*dims
    except:
        log.error(f"Error parsing data in a <rectangle>")
        return
    p0 = ll
    p1 = ll + dims

    # We're going to make a path so that we can use this with shape operations
    path = ET.SubElement(parent, 'path')
    diagram.add_id(path, element.get('id'))

    rotate = un.valid_eval(element.get('rotate', '0'))
    ctm = CTM.CTM()
    ctm.translate(*center)
    ctm.rotate(rotate)
    dx, dy = dims/2
    user_corners = [ctm.transform(p) for p in [
        (-dx, -dy), (dx, -dy), (dx, dy), (-dx, dy)
    ]
                    ]
    
    corners = [diagram.transform(c) for c in user_corners]

    radius = un.valid_eval(element.get('corner-radius', '0'))
    if radius == 0:
        cmds = ['M', util.pt2str(corners[0])]
        for c in corners[1:]:
            cmds += ['L', util.pt2str(c)]
        cmds.append('Z')
    else:
        cmds = []
        corners += corners[:2]
        for i in range(4):
            v1 = math_util.normalize(corners[i+1] - corners[i])
            v2 = math_util.normalize(corners[i+2] - corners[i+1])
            command = 'L'
            if len(cmds) == 0:
                command = 'M'
            cmds += [command, util.pt2str(corners[i+1] - radius*v1)]
            cmds += ['Q',
                     util.pt2str(corners[i+1]),
                     util.pt2str(corners[i+1] + radius*v2)]
        cmds.append('Z')

    path.set('d', ' '.join(cmds))

    if diagram.output_format() == 'tactile':
        if element.get('stroke') is not None:
            element.set('stroke', 'black')
        if element.get('fill') is not None:
            element.set('fill', 'lightgray')
    else:
        util.set_attr(element, 'stroke', 'none')
        util.set_attr(element, 'fill', 'none')

    util.set_attr(element, 'thickness', '2')
    util.add_attr(path, util.get_2d_attr(element))
    util.cliptobbox(path, element, diagram)

    if outline_status == 'add_outline':
        diagram.add_outline(element, path, parent)
        return

    if element.get('outline', 'no') == 'yes' or diagram.output_format() == 'tactile':
        diagram.add_outline(element, path, parent)
        finish_outline(element, diagram, parent)
    else:
        parent.append(path)

def finish_outline(element, diagram, parent):
    diagram.finish_outline(element,
                           element.get('stroke'),
                           element.get('thickness'),
                           element.get('fill', 'none'),
                           parent)

File: packages/prefig-rust/prefig-core/src/core/rectangle.rs
//! Port of prefig/core/rectangle.py.

use crate::core::ctm::CTM;
use crate::core::diagram::Diagram;
use crate::core::math_utilities::normalize;
use crate::core::utilities::{self as util, pt2str};
use crate::xml::{self, El};

pub fn rectangle(element: &El, diagram: &mut Diagram, parent: &El, outline_group: Option<&El>) {
    let ll_attr = element.borrow().get_or("lower-left", "(0,0)");
    let dims_attr = element.borrow().get_or("dimensions", "(1,1)");
    let center_attr = element.borrow().get("center");

    let eval_pair = |diagram: &mut Diagram, attr: &str| -> Option<[f64; 2]> {
        let v = diagram.ctx.valid_eval(attr).ok()?.as_vec_f64().ok()?;
        (v.len() >= 2).then(|| [v[0], v[1]])
    };
    let (Some(ll), Some(dims)) = (
        eval_pair(diagram, &ll_attr),
        eval_pair(diagram, &dims_attr),
    ) else {
        log::error!("Error parsing data in a <rectangle>");
        return;
    };
    let center = match center_attr {
        Some(attr) => {
            let Some(center) = eval_pair(diagram, &attr) else {
                log::error!("Error parsing data in a <rectangle>");
                return;
            };
            center
        }
        None => [ll[0] + 0.5 * dims[0], ll[1] + 0.5 * dims[1]],
    };

    let path = xml::sub_element(parent, "path");
    let id = element.borrow().get("id");
    diagram.add_id(&path, id.as_deref());
    diagram.register_svg_element(element, &path);

    let rotate_attr = element.borrow().get_or("rotate", "0");
    let rotate = diagram
        .ctx
        .valid_eval(&rotate_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(0.0);
    let mut ctm = CTM::new();
    ctm.translate(center[0], center[1]);
    ctm.rotate(rotate, true);
    let (dx, dy) = (dims[0] / 2.0, dims[1] / 2.0);
    let user_corners: Vec<[f64; 2]> = [[-dx, -dy], [dx, -dy], [dx, dy], [-dx, dy]]
        .iter()
        .map(|&p| ctm.transform(p))
        .collect();
    let mut corners: Vec<[f64; 2]> = user_corners
        .iter()
        .map(|&c| diagram.transform(c))
        .collect();

    let radius_attr = element.borrow().get_or("corner-radius", "0");
    let radius = diagram
        .ctx
        .valid_eval(&radius_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(0.0);
    let cmds = if radius == 0.0 {
        let mut cmds = vec!["M".to_string(), pt2str(corners[0], " ")];
        for c in &corners[1..] {
            cmds.push("L".to_string());
            cmds.push(pt2str(*c, " "));
        }
        cmds.push("Z".to_string());
        cmds
    } else {
        let mut cmds: Vec<String> = Vec::new();
        corners.push(corners[0]);
        corners.push(corners[1]);
        for i in 0..4 {
            let v1 = normalize([
                corners[i + 1][0] - corners[i][0],
                corners[i + 1][1] - corners[i][1],
            ]);
            let v2 = normalize([
                corners[i + 2][0] - corners[i + 1][0],
                corners[i + 2][1] - corners[i + 1][1],
            ]);
            let command = if cmds.is_empty() { "M" } else { "L" };
            cmds.push(command.to_string());
            cmds.push(pt2str(
                [
                    corners[i + 1][0] - radius * v1[0],
                    corners[i + 1][1] - radius * v1[1],
                ],
                " ",
            ));
            cmds.push("Q".to_string());
            cmds.push(pt2str(corners[i + 1], " "));
            cmds.push(pt2str(
                [
                    corners[i + 1][0] + radius * v2[0],
                    corners[i + 1][1] + radius * v2[1],
                ],
                " ",
            ));
        }
        cmds.push("Z".to_string());
        cmds
    };
    path.borrow_mut().set("d", &cmds.join(" "));

    if diagram.output_format() == "tactile" {
        let stroke = element.borrow().get("stroke");
        if stroke.is_some_and(|s| s != "none") {
            element.borrow_mut().set("stroke", "black");
        }
        util::set_tactile_fill(element);
    } else {
        util::set_attr(element, "stroke", "none", &mut diagram.ctx);
        util::set_attr(element, "fill", "none", &mut diagram.ctx);
    }

    util::set_attr(element, "thickness", "2", &mut diagram.ctx);
    let attrs = util::get_2d_attr(element, diagram);
    util::add_attr(&path, attrs);
    util::cliptobbox(&path, element, diagram);

    if let Some(outline_group) = outline_group {
        diagram.add_outline(element, &path, outline_group, None);
        finish_outline(element, diagram, parent);
    } else if element.borrow().get_or("outline", "no") == "yes"
        || diagram.output_format() == "tactile"
    {
        diagram.add_outline(element, &path, parent, None);
        finish_outline(element, diagram, parent);
    } else {
        xml::append(parent, &path);
    }
}

fn finish_outline(element: &El, diagram: &mut Diagram, parent: &El) {
    let stroke = element.borrow().get("stroke");
    let thickness = element.borrow().get("thickness");
    let fill = element.borrow().get_or("fill", "none");
    diagram.finish_outline(element, stroke, thickness, &fill, parent);
}
